File: dsa_leetcode_beginner/0881-boats-to-save-people/0881-boats-to-save-people.py
class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        people.sort()
        i, j = 0, len(people) - 1
        boats = 0
        
        # Loop until the two pointers meet or cross each other
        while i <= j: 
            boats += 1

            # Check if the lightest and heaviest person can share a boat
            if people[i] + people[j] <= limit:
                i += 1
            
            j -= 1

        # Return the total number of boats used
        return boats
    

#Question: https://leetcode.com/problems/boats-to-save-people
#Blog: https://blog.unwiredlearning.com/boats-to-save-people

File: dsa_leetcode_beginner/0881-boats-to-save-people/0881-boats-to-save-people.cpp
#include <algorithm>
#include <vector>

class Solution {
public:
    int numRescueBoats(std::vector<int>& people, int limit) {
        std::sort(people.begin(), people.end());
        int i = 0, j = people.size() - 1;
        int boats = 0;
        
        while (i <= j) {
            boats++;
            if (people[i] + people[j] <= limit) {
                i++;
            }
            j--;
        }
        
        return boats;
    }
};